File: getNextIndex.m
% function [curr_indexes,currx2,curry2] = getNextIndex(i,tot_weights,x,y,currx,curry,curr_indexes)
% 
% p = length(currx);
% n = size(tot_weights,2)-2;
% 
% W_normalized = tot_weights./repmat(sum(tot_weights,2),1,size(tot_weights,2));
% W_cdf = cumsum(W_normalized,2);
% rands = rand(size(tot_weights,1),1);
% C = rands<W_cdf;
% [~,curr_indexes] = max(C,[],2);
% 
% currx2 = nan(1,p);
% curry2 = nan(1,p);
% 
% currx2(curr_indexes==n+1) = 0;
% curry2(curr_indexes==n+1) = 0;
% 
% currx2(curr_indexes==n+3) = currx(curr_indexes==n+3);
% curry2(curr_indexes==n+3) = curry(curr_indexes==n+3);
% 
% curr_indexes(curr_indexes==n+1) = nan;
% curr_indexes(curr_indexes==n+2) = nan;
% curr_indexes(curr_indexes==n+3) = nan;
% 
% % indexes = [1:n,nan];
% % 
% % curr_indexes = nan(1,p);
% % 
% % for k = 1:p
% %     
% %     curr_indexes(k) = randsample(indexes,1,true,tot_weights(k,:));
% % 
% % end
% 
% % set currx and curry to all-nans of the correct size
% % update currx where curr_indexes is not nan
% % fill these with curr_indexes of x (excluding places where curr_indexes is nan)
% currx2(~isnan(curr_indexes)) = x(i,curr_indexes(~isnan(curr_indexes)));
% curry2(~isnan(curr_indexes)) = y(i,curr_indexes(~isnan(curr_indexes)));

function [curr_indexes,currx,curry] = getNextIndex(i,tot_weights,x,y,currx,curry,curr_indexes)

p = length(currx);
n = size(tot_weights,2)-2;

W_normalized = tot_weights./repmat(sum(tot_weights,2),1,size(tot_weights,2));
W_cdf = cumsum(W_normalized,2);
rands = rand(size(tot_weights,1),1);
C = rands<W_cdf;
[~,curr_indexes] = max(C,[],2);

currx = nan(1,p);
curry = nan(1,p);

currx(curr_indexes==n+1) = 0;
curry(curr_indexes==n+1) = 0;

curr_indexes(curr_indexes==n+1) = nan;
curr_indexes(curr_indexes==n+2) = nan;

% indexes = [1:n,nan];
% 
% curr_indexes = nan(1,p);
% 
% for k = 1:p
%     
%     curr_indexes(k) = randsample(indexes,1,true,tot_weights(k,:));
% 
% end

% set currx and curry to all-nans of the correct size
% update currx where curr_indexes is not nan
% fill these with curr_indexes of x (excluding places where curr_indexes is nan)
currx(~isnan(curr_indexes)) = x(i,curr_indexes(~isnan(curr_indexes)));
curry(~isnan(curr_indexes)) = y(i,curr_indexes(~isnan(curr_indexes)));

File: getNextIndex.py
import numpy as np

import numpy as np

def get_next_index(tot_weights, x, y, currx, curry, curr_indexes, L):
    """
    Translated from MATLAB getNextIndex.m
    Determines the next index and corresponding (x,y) position for each electron.
    """
    num_e, cols = tot_weights.shape # num_e is the number of electrons beign simulated, cols is 
    n = cols - 2  # number of particle targets (minus 2 electrodes)

    LEFT_ELECTRODE = n + 1
    RIGHT_ELECTRODE = n + 2

    # Normalize hopping probabilities
    row_sums = np.sum(tot_weights, axis=1, keepdims=True)
    W_normalized = tot_weights / row_sums
    W_cdf = np.cumsum(W_normalized, axis=1)

    # Random choice via CDF
    rands = np.random.rand(num_e, 1)
    C = rands < W_cdf
    next_indices = np.argmax(C, axis=1) + 1  # MATLAB is 1-based, Python is 0-based; adjust later if needed

    # Initialize next positions
    nextx = np.zeros(num_e)
    nexty = np.zeros(num_e)

    # Beginning electrode (x=0, y=nan)
    mask_left = next_indices == LEFT_ELECTRODE
    nextx[mask_left] = 0
    nexty[mask_left] = np.nan

    # Right electrode (x=L, y=nan)
    mask_right = next_indices == RIGHT_ELECTRODE
    nextx[mask_right] = L
    nexty[mask_right] = np.nan

    # Particles in between
    mask_particles = (~mask_left) & (~mask_right)
    nextx[mask_particles] = x[next_indices[mask_particles] - 1]
    nexty[mask_particles] = y[next_indices[mask_particles] - 1]

    return next_indices, nextx, nexty